src/main/scala/cognite/spark/v1/DataModelInstancesRelation.scala

```diff
@@ -291,92 +286,103 @@ object DataModelInstanceRelation {
+        case x: Iterable[Long] @unchecked if x.head.isInstanceOf[Long] =>
+          ArrayProperty(x.toVector.map(i => Float32Property(i.toFloat)))
+        case x: Iterable[java.math.BigDecimal] @unchecked if x.head.isInstanceOf[java.math.BigDecimal] =>
+          ArrayProperty(x.toVector.map(i => Float32Property(i.floatValue())))
+        case x: Iterable[java.math.BigInteger] @unchecked if x.head.isInstanceOf[java.math.BigInteger] =>
+          ArrayProperty(x.toVector.map(i => Float32Property(i.floatValue())))
+        case a => throw new CdfSparkException(notValidPropertyTypeMessage(a, propertyType))
+      }
+      case "float64[]" | "numeric[]" => {
+        case x: Iterable[_] if x.isEmpty => ArrayProperty(Vector.empty)
+        case x: Iterable[Double] @unchecked if x.head.isInstanceOf[Double] =>
+          ArrayProperty(x.toVector.map(Float64Property))
         case x: Iterable[Float] @unchecked if x.head.isInstanceOf[Float] =>
-          Json.fromValues(x.map(value => jsonFromDouble(value.doubleValue)))
+          ArrayProperty(x.toVector.map(f => Float64Property(f.toDouble)))
+        case x: Iterable[Int] @unchecked if x.head.isInstanceOf[Int] =>
+          ArrayProperty(x.toVector.map(f => Float64Property(f.toDouble)))
         case x: Iterable[Long] @unchecked if x.head.isInstanceOf[Long] =>
-          Json.fromValues(x.map(value => jsonFromDouble(value.doubleValue)))
+          ArrayProperty(x.toVector.map(f => Float64Property(f.toDouble)))
         case x: Iterable[java.math.BigDecimal] @unchecked if x.head.isInstanceOf[java.math.BigDecimal] =>
-          Json.fromValues(x.map(value => jsonFromDouble(value.doubleValue)))
+          ArrayProperty(x.toVector.map(i => Float64Property(i.doubleValue())))
         case x: Iterable[java.math.BigInteger] @unchecked if x.head.isInstanceOf[java.math.BigInteger] =>
-          Json.fromValues(x.map(value => jsonFromDouble(value.doubleValue)))
+          ArrayProperty(x.toVector.map(i => Float64Property(i.doubleValue())))
         case a => throw new CdfSparkException(notValidPropertyTypeMessage(a, propertyType))
       }
       case "boolean[]" => {
-        case null if !nullable =>
-          throw new CdfSparkException(propertyNotNullableMessage(propertyType)) // scalastyle:off null
-        case null => Json.Null // scalastyle:off null
-        case x: Iterable[_] if x.isEmpty => Json.arr()
+        case x: Iterable[_] if x.isEmpty => ArrayProperty(Vector.empty)
         case x: Iterable[Boolean] @unchecked if x.head.isInstanceOf[Boolean] =>
-          Json.fromValues(x.map(Json.fromBoolean))
+          ArrayProperty(x.toVector.map(BooleanProperty))
+        case a => throw new CdfSparkException(notValidPropertyTypeMessage(a, propertyType))
+      }
+      case "int[]" | "int32[]" => {
+        case x: Iterable[_] if x.isEmpty => ArrayProperty(Vector.empty)
+        case x: Iterable[Int] @unchecked if x.head.isInstanceOf[Int] =>
```

Comment:
I think you can just do `case x: Iterable[_] if x.head.isInstanceOf[Int]` here.

The `@unchecked` is needed because due to type erasure it's not truly possible to match on `Iterable[Int]`, at runtime it's only `Iterable` with no type information stored for the element types, so the type check against `Iterable[Int]` is a lie, and that's why the compiler requires us to mark it as `@unchecked` (as it can't be checked).

Reply:
It's not about IntelliJ's type matching, it's about the JVM and type erasure.
At runtime, the type of `x` will be `Iterable`. It's not `Iterable[Int]`, `Iterable[Double]`, etc. it's compiled into only `Iterable`. Type parameters don't exist in JVM bytecode.

The Scala compiler knows this, and it knows that it can't actually check if `x` is of type `Iterable[Int]`. That's why it fails to compile. Adding `@unchecked` says "trust me, I know what I'm doing" and in our case we know what we're doing (we think) because of the `x.head.isInstanceOf[Int]` check.